### 1. Paired end 

- config/config_paired.yaml

```yaml

###################### Sample info ######################

SAMPLE:
  SRR8245176: Untreated1
  SRR8245177: Untreated2
  SRR8245178: Untreated3
  SRR8245179: Treated1
  SRR8245180: Treated2
  SRR8245181: Treated3
  




END: 
  - 1
  - 2


SRA:
  - SRR8245176
  - SRR8245177
  - SRR8245178
  - SRR8245179
  - SRR8245180
  - SRR8245181
  

FASTQ_PREFIX:
  - Untreated1
  - Untreated2
  - Untreated3
  - Treated1
  - Treated2
  - Treated3
  


###################### Reference info ######################
  
REFERENCE_LINK:
  GENOME: 
    - 'ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/GRCh38.primary_assembly.genome.fa.gz'
    - 'GRCh38.primary_assembly.genome.fa.gz'
    - 'GRCh38.primary_assembly.genome.fa'
  ANNOTATION: 
    - 'ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_37/gencode.v37.primary_assembly.annotation.gtf.gz'
    - 'gencode.v37.primary_assembly.annotation.gtf.gz'
    - 'gencode.v37.primary_assembly.annotation.gtf'


###################### Extra-setting info ######################

INDEX_STAR: "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index" # Assigns path to star indexing directory (ABSOLUTE PATH NEEDED due to a potential STAR error!)

INDEX_HISAT: "reference/hisat2_index/hisat2_index" # Assigns hisat2 index files (e.g.reference/hisat2_index/hisat2_index.1.ht2, reference/hisat2_index/hisat2_index.2.ht2, ...)


FASTQ_EXT: '.fastq.gz'

CONDA: "config/conda_r.yaml"

RCONFIG: 'config/config_single.R'
  

  
```

- Snakefile

```

#################################### Defined by users #################################
configfile: "config/config_paired.yaml"    # Sets path to the config file
#######################################################################################

shell.prefix('set -euo pipefail; ')
shell.executable('/bin/bash')



rule all: 
    input: 
        "DE_analysis/DE.html"

rule get_fastq:   # Creates fastq.gz files in fastq directory
    """
    This rule downloads SRA and converts to FASTQ files
    """
    output:
        expand("fastq/{sample}_{end}.fastq.gz", sample=config['FASTQ_PREFIX'], end=config['END'])  # Gzipped FASTQ files from SRA 
    params:
        dic=config['SAMPLE'],   # Sample dictionary
        reads=config['END'],    # Reads (e.g. [1] or [1, 2]) 
        sra=config['SRA']       # SRA number 
    run:
        shell("fastq-dump --split-files {params.sra} --gzip")    # with or without -X 100000
        for key, value in params.dic.items(): 
              for read in params.reads: 
                  shell("mv {key}_{read}.fastq.gz fastq/{value}_{read}.fastq.gz") 


rule get_reference:    
    """
    This rule downloads reference files
    """
    params:
        gen_link=config['REFERENCE_LINK']['GENOME'][0],   # Gencode reference genome file link 
        gen_name=config['REFERENCE_LINK']['GENOME'][1],   # Output reference genome location & name 
        anno_link=config['REFERENCE_LINK']['ANNOTATION'][0],  # Gencode GTF (annotation) file link
        anno_name=config['REFERENCE_LINK']['ANNOTATION'][1]   # Output GTF file location & name
    output:
        gen=expand("reference/{gen}", gen=config['REFERENCE_LINK']['GENOME'][2]),  # Decompressed reference genome file 
        anno=expand("reference/{anno}", anno=config['REFERENCE_LINK']['ANNOTATION'][2])  # Decompressed GTF file
    shell:
        "set +o pipefail; "
        "wget -c {params.gen_link} -O reference/{params.gen_name} && "
        "wget -c {params.anno_link} -O reference/{params.anno_name} && "
        "gzip -d reference/*.gz"

rule index_hisat2:
    """
    This rule constructs HISAT2 index files
    """
    input: 
        expand("reference/{gen}", gen=config['REFERENCE_LINK']['GENOME'][2])    # Decompressed reference genome file
    output:
        expand("reference/hisat2_index/hisat2_index.{number}.ht2", number=[x+1 for x in range(8)])   # HISAT2 indexing files
    threads: 8
    shell:
        "set +o pipefail; "
        "hisat2-build -f -o 4 "
        "-p {threads} "
        "--seed 67 "
        "{input} "
        "hisat2_index && "
        "mv *.ht2 reference/hisat2_index"


rule index_star:
    """
    This rule constructs STAR index files
    """
    input:
        fa=expand("reference/{gen}", gen=config['REFERENCE_LINK']['GENOME'][2]),  # Decompressed reference genome file
        gtf=expand("reference/{anno}", anno=config['REFERENCE_LINK']['ANNOTATION'][2])  # Decompressed GTF file
    output:
        "reference/star_index/Genome",   # STAR indexing files
        "reference/star_index/SA",       # STAR indexing files
        "reference/star_index/SAindex"   # STAR indexing files
    threads: 10
    shell:
        "set +o pipefail; "
        "STAR --runThreadN {threads} "
        "--runMode genomeGenerate "
        "--genomeDir reference/star_index "
        "--genomeFastaFiles {input.fa} "
        "--sjdbGTFfile {input.gtf}"




rule align_hisat2:    # Creates bam files in hisat2_output directory"
    """
    This rule aligns the reads using HISAT2    
    """
    input:
        fastq=expand("fastq/{sample}_{end}.fastq.gz", sample=config['FASTQ_PREFIX'], end=config['END']),  # Gzipped FASTQ files
        index=expand("reference/hisat2_index/hisat2_index.{number}.ht2", number=[x+1 for x in range(8)])  # HISAT2 indexing files
    output:
        expand("hisat2_output/{sample}.bam", sample=config['FASTQ_PREFIX']),   # Bam files
        temp(expand("hisat2_output/{sample}.sam", sample=config['FASTQ_PREFIX']))
    params:
        files=config["FASTQ_PREFIX"],  # e.g. Ctrl, Treatment
        ext=config['FASTQ_EXT'],       # extension of the FASTQ files (e.g. .fastq.gz)
        indexing=config['INDEX_HISAT'] # HISAT2 indexing location and file name prefix
    threads: 8
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1= "fastq/" + params.files[i] + "_1" + params.ext 
            r2=""
            read="-U " + r1
            if len(input.fastq) == 2 * len(params.files): 
                r2= "fastq/" + params.files[i] + "_2" + params.ext  
                read="-1 " + r1 + " -2 " + r2
            shell("hisat2 -q -p {threads} "
                  "--seed 23 "
                  "--summary-file hisat2_output/summary_{p}.txt "
                  "-x {params.indexing} "
                  "{read} "
                  "-S hisat2_output/{p}.sam && "
                  "samtools view -bS "
                  "-@ {threads} "
                  "hisat2_output/{p}.sam > hisat2_output/{p}.bam")



rule align_star:   # Creates bam files in star_output directory"
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=expand("reference/{gen}", gen=config['REFERENCE_LINK']['ANNOTATION'][2]),  # Decompressed GTF file
        fastq=expand("fastq/{sample}_{end}.fastq.gz", sample=config['FASTQ_PREFIX'], end=config['END']),                  # Gzipped FASTQ files
        index1="reference/star_index/Genome",
        index2="reference/star_index/SA",
        index3="reference/star_index/SAindex"
    output:
        expand("star_output/{sample}Aligned.sortedByCoord.out.bam", sample=config['FASTQ_PREFIX'])  # Bam files
    params:
        indexing=config["INDEX_STAR"],  # STAR indexing file directory
        files=config["FASTQ_PREFIX"],   # e.g. Ctrl, Treatment
        ext=config['FASTQ_EXT']         # extension of the FASTQ files (e.g. fastq.gz)
    threads: 10
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1= "fastq/" + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(input.fastq) == 2 * len(params.files): 
                r2= "fastq/" + params.files[i] + "_2" + params.ext + " " 
            shell("STAR --runThreadN {threads} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile {input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix star_output/{p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions")     



rule featurecounts:
    """
    This rule assesses read counts using featureCounts
    """
    input:
        star=expand("star_output/{sample}Aligned.sortedByCoord.out.bam", sample=config['FASTQ_PREFIX']),  # Bam files from STAR
        hisat2=expand("hisat2_output/{sample}.bam", sample=config['FASTQ_PREFIX'])    # Bam files from HISAT2
    params:
        ends=config['END'], # Read ends (e.g. [1] or [1, 2])
        gtf=expand("reference/{anno}", anno=config['REFERENCE_LINK']['ANNOTATION'][2])  # Decompressed GTF file
    output:
        star="star_output/featurecounts.tsv",  # Read count tsv file (STAR version)
        hisat2="hisat2_output/featurecounts.tsv"  # Read count tsv file (HISAT2 version)
    threads: 16
    run:
        count_star=output.star
        count_hisat2=output.hisat2
        bam_star=input.star
        bam_hisat2=input.hisat2
        end=""
        if len(params.ends) == 2:
            end='-p'
        # Additional featureCounts flags
        # -F: format of the annotation file. 'GTF' by default.
        # -g: attribute type. 'gene_id' by default. 
        # -L: set for long-read inputs 
        # -s: strand-specificity. 0 (unstranded & default), 1 (stranded), 2 (reversely stranded)
        shell("set +o pipefail; "
              "featureCounts {end} "    
              "-s0 "
              "-T {threads} "
              "-a {params.gtf} "
              "-o {count_star} "
              "{bam_star} &> {count_star}.log && "
              "featureCounts {end} "    
              "-s0 "
              "-T {threads} "
              "-a {params.gtf} "
              "-o {count_hisat2} "
              "{bam_hisat2} &> {count_hisat2}.log")



rule deseq2:
    """
    This rule performs differential expression (DE) analysis using DESeq2 in R
    """
    input:
        star_reads="star_output/featurecounts.tsv",  # Assessed read counts from STAR aligner 
        hisat2_reads="hisat2_output/featurecounts.tsv",  # Assessed read counts from HISAT2 aligner
        rmd="DE_analysis/DE.Rmd",                    # R scripts for downstream DE analysis
        rconfig=config["RCONFIG"]                    # config file for Rmd 
    output:
        "DE_analysis/DE.html",           # Output html file from Rmd 
        "DE_analysis/lfc_STAR.csv",      # DESeq2 result from STAR-mediated alignment 
        "DE_analysis/lfc_HISAT2.csv",    # DESeq2 result from HISAT2-mediated alignment
        "DE_analysis/baseMean_difference.csv",        # Gene discordance table in baseMean
        "DE_analysis/log2FoldChange_difference.csv",  # Gene discordance table in log2FoldChange
        "DE_analysis/padj_difference.csv",            # Gene discordance table in padj
        temp("DE_analysis/trf_input.fa") 
    conda:
        config['CONDA']                           # conda env config file
    shell:
        "Rscript -e \"rmarkdown::render('{input.rmd}')\""   # Requires "snakemake -j 8 --use-conda" command when running snakemake

```

- Errors: Bad Request (HTTP 400)

    - Reference: [API Error Codes](https://docs.rstudio.com/connect/1.7.2/api/#API-Error-Codes)

```


(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8 --use-conda
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	align_hisat2
	1	align_star
	1	all
	1	deseq2
	1	featurecounts
	1	get_fastq
	6
Select jobs to execute...

[Tue May 11 20:13:54 2021]
rule get_fastq:
    output: fastq/Untreated1_1.fastq.gz, fastq/Untreated1_2.fastq.gz, fastq/Untreated2_1.fastq.gz, fastq/Untreated2_2.fastq.gz, fastq/Untreated3_1.fastq.gz, fastq/Untreated3_2.fastq.gz, fastq/Treated1_1.fastq.gz, fastq/Treated1_2.fastq.gz, fastq/Treated2_1.fastq.gz, fastq/Treated2_2.fastq.gz, fastq/Treated3_1.fastq.gz, fastq/Treated3_2.fastq.gz
    jobid: 5

Job counts:
	count	jobs
	1	get_fastq
	1
Read 12928713 spots for SRR8245176
Written 12928713 spots for SRR8245176
Read 12525467 spots for SRR8245177
Written 12525467 spots for SRR8245177
Read 12145797 spots for SRR8245178
Written 12145797 spots for SRR8245178
Read 12182934 spots for SRR8245179
Written 12182934 spots for SRR8245179
Read 12241191 spots for SRR8245180
Written 12241191 spots for SRR8245180
Read 12974834 spots for SRR8245181
Written 12974834 spots for SRR8245181
Read 74998936 spots total
Written 74998936 spots total
[Tue May 11 21:52:18 2021]
Finished job 5.
1 of 6 steps (17%) done
Select jobs to execute...

[Tue May 11 21:52:18 2021]
rule align_hisat2:
    input: fastq/Untreated1_1.fastq.gz, fastq/Untreated1_2.fastq.gz, fastq/Untreated2_1.fastq.gz, fastq/Untreated2_2.fastq.gz, fastq/Untreated3_1.fastq.gz, fastq/Untreated3_2.fastq.gz, fastq/Treated1_1.fastq.gz, fastq/Treated1_2.fastq.gz, fastq/Treated2_1.fastq.gz, fastq/Treated2_2.fastq.gz, fastq/Treated3_1.fastq.gz, fastq/Treated3_2.fastq.gz, reference/hisat2_index/hisat2_index.1.ht2, reference/hisat2_index/hisat2_index.2.ht2, reference/hisat2_index/hisat2_index.3.ht2, reference/hisat2_index/hisat2_index.4.ht2, reference/hisat2_index/hisat2_index.5.ht2, reference/hisat2_index/hisat2_index.6.ht2, reference/hisat2_index/hisat2_index.7.ht2, reference/hisat2_index/hisat2_index.8.ht2
    output: hisat2_output/Untreated1.bam, hisat2_output/Untreated2.bam, hisat2_output/Untreated3.bam, hisat2_output/Treated1.bam, hisat2_output/Treated2.bam, hisat2_output/Treated3.bam, hisat2_output/Untreated1.sam, hisat2_output/Untreated2.sam, hisat2_output/Untreated3.sam, hisat2_output/Treated1.sam, hisat2_output/Treated2.sam, hisat2_output/Treated3.sam
    jobid: 7
    threads: 8

Job counts:
	count	jobs
	1	align_hisat2
	1
12928713 reads; of these:
  12928713 (100.00%) were paired; of these:
    1022402 (7.91%) aligned concordantly 0 times
    11161764 (86.33%) aligned concordantly exactly 1 time
    744547 (5.76%) aligned concordantly >1 times
    ----
    1022402 pairs aligned concordantly 0 times; of these:
      271482 (26.55%) aligned discordantly 1 time
    ----
    750920 pairs aligned 0 times concordantly or discordantly; of these:
      1501840 mates make up the pairs; of these:
        1039641 (69.22%) aligned 0 times
        408049 (27.17%) aligned exactly 1 time
        54150 (3.61%) aligned >1 times
95.98% overall alignment rate
12525467 reads; of these:
  12525467 (100.00%) were paired; of these:
    967756 (7.73%) aligned concordantly 0 times
    11005537 (87.87%) aligned concordantly exactly 1 time
    552174 (4.41%) aligned concordantly >1 times
    ----
    967756 pairs aligned concordantly 0 times; of these:
      263711 (27.25%) aligned discordantly 1 time
    ----
    704045 pairs aligned 0 times concordantly or discordantly; of these:
      1408090 mates make up the pairs; of these:
        946349 (67.21%) aligned 0 times
        412547 (29.30%) aligned exactly 1 time
        49194 (3.49%) aligned >1 times
96.22% overall alignment rate
12145797 reads; of these:
  12145797 (100.00%) were paired; of these:
    1177291 (9.69%) aligned concordantly 0 times
    10444399 (85.99%) aligned concordantly exactly 1 time
    524107 (4.32%) aligned concordantly >1 times
    ----
    1177291 pairs aligned concordantly 0 times; of these:
      313058 (26.59%) aligned discordantly 1 time
    ----
    864233 pairs aligned 0 times concordantly or discordantly; of these:
      1728466 mates make up the pairs; of these:
        1241341 (71.82%) aligned 0 times
        432831 (25.04%) aligned exactly 1 time
        54294 (3.14%) aligned >1 times
94.89% overall alignment rate
12182934 reads; of these:
  12182934 (100.00%) were paired; of these:
    948136 (7.78%) aligned concordantly 0 times
    10642842 (87.36%) aligned concordantly exactly 1 time
    591956 (4.86%) aligned concordantly >1 times
    ----
    948136 pairs aligned concordantly 0 times; of these:
      245882 (25.93%) aligned discordantly 1 time
    ----
    702254 pairs aligned 0 times concordantly or discordantly; of these:
      1404508 mates make up the pairs; of these:
        1007297 (71.72%) aligned 0 times
        349987 (24.92%) aligned exactly 1 time
        47224 (3.36%) aligned >1 times
95.87% overall alignment rate
12241191 reads; of these:
  12241191 (100.00%) were paired; of these:
    1038454 (8.48%) aligned concordantly 0 times
    10599308 (86.59%) aligned concordantly exactly 1 time
    603429 (4.93%) aligned concordantly >1 times
    ----
    1038454 pairs aligned concordantly 0 times; of these:
      261588 (25.19%) aligned discordantly 1 time
    ----
    776866 pairs aligned 0 times concordantly or discordantly; of these:
      1553732 mates make up the pairs; of these:
        1132045 (72.86%) aligned 0 times
        370531 (23.85%) aligned exactly 1 time
        51156 (3.29%) aligned >1 times
95.38% overall alignment rate
12974834 reads; of these:
  12974834 (100.00%) were paired; of these:
    966195 (7.45%) aligned concordantly 0 times
    11409721 (87.94%) aligned concordantly exactly 1 time
    598918 (4.62%) aligned concordantly >1 times
    ----
    966195 pairs aligned concordantly 0 times; of these:
      248349 (25.70%) aligned discordantly 1 time
    ----
    717846 pairs aligned 0 times concordantly or discordantly; of these:
      1435692 mates make up the pairs; of these:
        1005630 (70.04%) aligned 0 times
        378947 (26.39%) aligned exactly 1 time
        51115 (3.56%) aligned >1 times
96.12% overall alignment rate
Removing temporary output file hisat2_output/Untreated1.sam.
Removing temporary output file hisat2_output/Untreated2.sam.
Removing temporary output file hisat2_output/Untreated3.sam.
Removing temporary output file hisat2_output/Treated1.sam.
Removing temporary output file hisat2_output/Treated2.sam.
Removing temporary output file hisat2_output/Treated3.sam.
[Tue May 11 22:09:27 2021]
Finished job 7.
2 of 6 steps (33%) done
Select jobs to execute...

[Tue May 11 22:09:27 2021]
rule align_star:
    input: reference/gencode.v37.primary_assembly.annotation.gtf, fastq/Untreated1_1.fastq.gz, fastq/Untreated1_2.fastq.gz, fastq/Untreated2_1.fastq.gz, fastq/Untreated2_2.fastq.gz, fastq/Untreated3_1.fastq.gz, fastq/Untreated3_2.fastq.gz, fastq/Treated1_1.fastq.gz, fastq/Treated1_2.fastq.gz, fastq/Treated2_1.fastq.gz, fastq/Treated2_2.fastq.gz, fastq/Treated3_1.fastq.gz, fastq/Treated3_2.fastq.gz, reference/star_index/Genome, reference/star_index/SA, reference/star_index/SAindex
    output: star_output/Untreated1Aligned.sortedByCoord.out.bam, star_output/Untreated2Aligned.sortedByCoord.out.bam, star_output/Untreated3Aligned.sortedByCoord.out.bam, star_output/Treated1Aligned.sortedByCoord.out.bam, star_output/Treated2Aligned.sortedByCoord.out.bam, star_output/Treated3Aligned.sortedByCoord.out.bam
    jobid: 3
    threads: 8

Job counts:
	count	jobs
	1	align_star
	1
May 11 22:09:28 ..... started STAR run
May 11 22:09:28 ..... loading genome
May 11 22:09:53 ..... processing annotations GTF
May 11 22:10:04 ..... inserting junctions into the genome indices
May 11 22:10:56 ..... started 1st pass mapping
May 11 22:12:34 ..... finished 1st pass mapping
May 11 22:12:34 ..... inserting junctions into the genome indices
May 11 22:13:39 ..... started mapping
May 11 22:15:34 ..... finished mapping
May 11 22:15:36 ..... started sorting BAM
May 11 22:15:49 ..... finished successfully
May 11 22:15:50 ..... started STAR run
May 11 22:15:50 ..... loading genome
May 11 22:16:14 ..... processing annotations GTF
May 11 22:16:25 ..... inserting junctions into the genome indices
May 11 22:17:17 ..... started 1st pass mapping
May 11 22:18:49 ..... finished 1st pass mapping
May 11 22:18:49 ..... inserting junctions into the genome indices
May 11 22:19:55 ..... started mapping
May 11 22:21:46 ..... finished mapping
May 11 22:21:47 ..... started sorting BAM
May 11 22:22:00 ..... finished successfully
May 11 22:22:00 ..... started STAR run
May 11 22:22:00 ..... loading genome
May 11 22:22:26 ..... processing annotations GTF
May 11 22:22:37 ..... inserting junctions into the genome indices
May 11 22:23:29 ..... started 1st pass mapping
May 11 22:25:01 ..... finished 1st pass mapping
May 11 22:25:02 ..... inserting junctions into the genome indices
May 11 22:26:07 ..... started mapping
May 11 22:27:55 ..... finished mapping
May 11 22:27:57 ..... started sorting BAM
May 11 22:28:08 ..... finished successfully
May 11 22:28:09 ..... started STAR run
May 11 22:28:09 ..... loading genome
May 11 22:28:35 ..... processing annotations GTF
May 11 22:28:45 ..... inserting junctions into the genome indices
May 11 22:29:39 ..... started 1st pass mapping
May 11 22:31:08 ..... finished 1st pass mapping
May 11 22:31:08 ..... inserting junctions into the genome indices
May 11 22:32:12 ..... started mapping
May 11 22:34:10 ..... finished mapping
May 11 22:34:11 ..... started sorting BAM
May 11 22:34:23 ..... finished successfully
May 11 22:34:24 ..... started STAR run
May 11 22:34:24 ..... loading genome
May 11 22:34:44 ..... processing annotations GTF
May 11 22:34:54 ..... inserting junctions into the genome indices
May 11 22:35:48 ..... started 1st pass mapping
May 11 22:37:17 ..... finished 1st pass mapping
May 11 22:37:17 ..... inserting junctions into the genome indices
May 11 22:38:22 ..... started mapping
May 11 22:40:21 ..... finished mapping
May 11 22:40:22 ..... started sorting BAM
May 11 22:40:35 ..... finished successfully
May 11 22:40:35 ..... started STAR run
May 11 22:40:35 ..... loading genome
May 11 22:40:51 ..... processing annotations GTF
May 11 22:41:01 ..... inserting junctions into the genome indices
May 11 22:41:54 ..... started 1st pass mapping
May 11 22:43:26 ..... finished 1st pass mapping
May 11 22:43:27 ..... inserting junctions into the genome indices
May 11 22:44:33 ..... started mapping
May 11 22:46:43 ..... finished mapping
May 11 22:46:44 ..... started sorting BAM
May 11 22:46:57 ..... finished successfully
[Tue May 11 22:46:58 2021]
Finished job 3.
3 of 6 steps (50%) done
Select jobs to execute...

[Tue May 11 22:46:58 2021]
rule featurecounts:
    input: star_output/Untreated1Aligned.sortedByCoord.out.bam, star_output/Untreated2Aligned.sortedByCoord.out.bam, star_output/Untreated3Aligned.sortedByCoord.out.bam, star_output/Treated1Aligned.sortedByCoord.out.bam, star_output/Treated2Aligned.sortedByCoord.out.bam, star_output/Treated3Aligned.sortedByCoord.out.bam, hisat2_output/Untreated1.bam, hisat2_output/Untreated2.bam, hisat2_output/Untreated3.bam, hisat2_output/Treated1.bam, hisat2_output/Treated2.bam, hisat2_output/Treated3.bam
    output: star_output/featurecounts.tsv, hisat2_output/featurecounts.tsv
    jobid: 2
    threads: 8

Job counts:
	count	jobs
	1	featurecounts
	1
[Tue May 11 22:48:02 2021]
Finished job 2.
4 of 6 steps (67%) done
Select jobs to execute...

[Tue May 11 22:48:02 2021]
rule deseq2:
    input: star_output/featurecounts.tsv, hisat2_output/featurecounts.tsv, DE_analysis/DE.Rmd, config/config_single.R
    output: DE_analysis/DE.html, DE_analysis/lfc_STAR.csv, DE_analysis/lfc_HISAT2.csv, DE_analysis/baseMean_difference.csv, DE_analysis/log2FoldChange_difference.csv, DE_analysis/padj_difference.csv, DE_analysis/trf_input.fa
    jobid: 1

Activating conda environment: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/conda/f66a882151a23b1708e7d8b19c2fee6c


processing file: DE.Rmd
  |.                                                                     |   1%
  ordinary text without R code

  |..                                                                    |   3%
label: global_options (with options) 
List of 1
 $ include: logi FALSE

  |...                                                                   |   4%
  ordinary text without R code

  |....                                                                  |   6%
label: loading_packages
── Attaching packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse 1.3.1 ──
✔ ggplot2 3.3.3     ✔ purrr   0.3.4
✔ tibble  3.1.1     ✔ dplyr   1.0.6
✔ tidyr   1.1.3     ✔ stringr 1.4.0
✔ readr   1.4.0     ✔ forcats 0.5.1
── Conflicts ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::between()   masks data.table::between()
✖ dplyr::filter()    masks stats::filter()
✖ dplyr::first()     masks data.table::first()
✖ dplyr::lag()       masks stats::lag()
✖ dplyr::last()      masks data.table::last()
✖ purrr::transpose() masks data.table::transpose()
Loading required package: BiocGenerics
Loading required package: parallel

Attaching package: 'BiocGenerics'

The following objects are masked from 'package:parallel':

    clusterApply, clusterApplyLB, clusterCall, clusterEvalQ,
    clusterExport, clusterMap, parApply, parCapply, parLapply,
    parLapplyLB, parRapply, parSapply, parSapplyLB

The following objects are masked from 'package:dplyr':

    combine, intersect, setdiff, union

The following objects are masked from 'package:stats':

    IQR, mad, sd, var, xtabs

The following objects are masked from 'package:base':

    anyDuplicated, append, as.data.frame, basename, cbind, colnames,
    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,
    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,
    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,
    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,
    union, unique, unsplit, which.max, which.min

Loading required package: BiocFileCache
Loading required package: dbplyr

Attaching package: 'dbplyr'

The following objects are masked from 'package:dplyr':

    ident, sql

Loading required package: S4Vectors
Loading required package: stats4

Attaching package: 'S4Vectors'

The following objects are masked from 'package:dplyr':

    first, rename

The following object is masked from 'package:tidyr':

    expand

The following objects are masked from 'package:data.table':

    first, second

The following object is masked from 'package:base':

    expand.grid

Loading required package: IRanges

Attaching package: 'IRanges'

The following objects are masked from 'package:dplyr':

    collapse, desc, slice

The following object is masked from 'package:purrr':

    reduce

The following object is masked from 'package:data.table':

    shift

Loading required package: GenomicRanges
Loading required package: GenomeInfoDb
Loading required package: SummarizedExperiment
Loading required package: MatrixGenerics
Loading required package: matrixStats

Attaching package: 'matrixStats'

The following object is masked from 'package:dplyr':

    count


Attaching package: 'MatrixGenerics'

The following objects are masked from 'package:matrixStats':

    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,
    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,
    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,
    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,
    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,
    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,
    colWeightedMeans, colWeightedMedians, colWeightedSds,
    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,
    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,
    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,
    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,
    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,
    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,
    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,
    rowWeightedSds, rowWeightedVars

Loading required package: Biobase
Welcome to Bioconductor

    Vignettes contain introductory material; view with
    'browseVignettes()'. To cite Bioconductor, see
    'citation("Biobase")', and for packages 'citation("pkgname")'.


Attaching package: 'Biobase'

The following object is masked from 'package:MatrixGenerics':

    rowMedians

The following objects are masked from 'package:matrixStats':

    anyMissing, rowMedians

The following object is masked from 'package:AnnotationHub':

    cache

Loading required package: GenomicFeatures
Loading required package: AnnotationDbi

Attaching package: 'AnnotationDbi'

The following object is masked from 'package:dplyr':

    select

Loading required package: AnnotationFilter

Attaching package: 'ensembldb'

The following object is masked from 'package:dplyr':

    filter

The following object is masked from 'package:stats':

    filter


Attaching package: 'gridExtra'

The following object is masked from 'package:Biobase':

    combine

The following object is masked from 'package:BiocGenerics':

    combine

The following object is masked from 'package:dplyr':

    combine


Attaching package: 'httr'

The following object is masked from 'package:Biobase':

    content


Attaching package: 'jsonlite'

The following object is masked from 'package:purrr':

    flatten

  |.....                                                                 |   7%
  ordinary text without R code

  |......                                                                |   9%
label: unnamed-chunk-1
  |.......                                                               |  10%
  ordinary text without R code

  |........                                                              |  12%
label: annotationhub
snapshotDate(): 2020-10-27
loading from cache
  |.........                                                             |  13%
  ordinary text without R code

  |..........                                                            |  14%
label: read_countmatrix
  |...........                                                           |  16%
  ordinary text without R code

  |............                                                          |  17%
label: generating_metadata
  |.............                                                         |  19%
  ordinary text without R code

  |..............                                                        |  20%
label: count_dataframe_cleaning
  |...............                                                       |  22%
  ordinary text without R code

  |................                                                      |  23%
label: library_size
  |.................                                                     |  25%
  ordinary text without R code

  |..................                                                    |  26%
label: generating_deseq2_objects
  |...................                                                   |  28%
  ordinary text without R code

  |....................                                                  |  29%
label: size_factors
  |.....................                                                 |  30%
  ordinary text without R code

  |......................                                                |  32%
label: dispersion_waldtest
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
  |.......................                                               |  33%
  ordinary text without R code

  |........................                                              |  35%
label: QC_PCA
  |.........................                                             |  36%
  ordinary text without R code

  |..........................                                            |  38%
label: QC_correlation_heatmap
  |...........................                                           |  39%
  ordinary text without R code

  |............................                                          |  41%
label: DE_analysis
using pre-existing size factors
estimating dispersions
found already estimated dispersions, replacing these
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
using pre-existing size factors
estimating dispersions
found already estimated dispersions, replacing these
gene-wise dispersion estimates
mean-dispersion relationship
final dispersion estimates
fitting model and testing
  |.............................                                         |  42%
  ordinary text without R code

  |..............................                                        |  43%
label: dispersion_plot
  |...............................                                       |  45%
  ordinary text without R code

  |................................                                      |  46%
label: setting_resultcondition
  |.................................                                     |  48%
  ordinary text without R code

  |..................................                                    |  49%
label: DEresult_extraction
  |....................................                                  |  51%
  ordinary text without R code

  |.....................................                                 |  52%
label: FDR_distribution
Scale for 'x' is already present. Adding another scale for 'x', which will
replace the existing scale.
  |......................................                                |  54%
  ordinary text without R code

  |.......................................                               |  55%
label: L2FC_distribution
  |........................................                              |  57%
  ordinary text without R code

  |.........................................                             |  58%
label: MAplot
  |..........................................                            |  59%
  ordinary text without R code

  |...........................................                           |  61%
label: expression_heatmap
  |............................................                          |  62%
  ordinary text without R code

  |.............................................                         |  64%
label: NA_genes
  |..............................................                        |  65%
  ordinary text without R code

  |...............................................                       |  67%
label: diff_prep
  |................................................                      |  68%
  ordinary text without R code

  |.................................................                     |  70%
label: comparison (with options) 
List of 1
 $ fig_width: num 12

  |..................................................                    |  71%
  ordinary text without R code

  |...................................................                   |  72%
label: distribution (with options) 
List of 2
 $ fig.height: num 12
 $ fig.width : num 12

  |....................................................                  |  74%
  ordinary text without R code

  |.....................................................                 |  75%
label: percent_difference (with options) 
List of 1
 $ fig.width: num 12

  |......................................................                |  77%
  ordinary text without R code

  |.......................................................               |  78%
label: ranking_comparison_prep
  |........................................................              |  80%
  ordinary text without R code

  |.........................................................             |  81%
label: ranking_comparison (with options) 
List of 1
 $ fig.width: num 12

  |..........................................................            |  83%
  ordinary text without R code

  |...........................................................           |  84%
label: saving_diff_csv
  |............................................................          |  86%
  ordinary text without R code

  |.............................................................         |  87%
label: count_repetitive_elements
Quitting from lines 1336-1484 (DE.Rmd) 
Error in eval(expr, envir, enclos) : Bad Request (HTTP 400).
Calls: <Anonymous> ... withCallingHandlers -> withVisible -> eval -> eval -> stop_for_status
In addition: There were 11 warnings (use warnings() to see them)

Execution halted
[Tue May 11 23:18:57 2021]
Error in rule deseq2:
    jobid: 1
    output: DE_analysis/DE.html, DE_analysis/lfc_STAR.csv, DE_analysis/lfc_HISAT2.csv, DE_analysis/baseMean_difference.csv, DE_analysis/log2FoldChange_difference.csv, DE_analysis/padj_difference.csv, DE_analysis/trf_input.fa
    conda-env: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/conda/f66a882151a23b1708e7d8b19c2fee6c
    shell:
        Rscript -e "rmarkdown::render('DE_analysis/DE.Rmd')"
        (one of the commands exited with non-zero exit code; note that snakemake uses bash strict mode!)

Removing output files of failed job deseq2 since they might be corrupted:
DE_analysis/lfc_STAR.csv, DE_analysis/lfc_HISAT2.csv, DE_analysis/baseMean_difference.csv, DE_analysis/log2FoldChange_difference.csv, DE_analysis/padj_difference.csv, DE_analysis/trf_input.fa
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/log/2021-05-11T201354.177614.snakemake.log

```
