### 1. Paired end 

- config/config_paired.yaml

```

GTF: "reference/gencode.v36.primary_assembly.annotation.gtf" # Assigns path to .gtf file

INDEX_STAR: "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index" # Assigns path to star indexing directory (ABSOLUTE PATH NEEDED due to a potential STAR error!)

INDEX_HISAT: "reference/hisat2_index" # Assigns path to hisat2 indexing directory

THREADS: 8     # Assigns the number of threads

STAR_OUT: "star_output"  # Assigns STAR output directory

HISAT2_OUT: "hisat2_output"  # Assigns HISAT2 output directory

INPUT_DIR: 'fastq_pe/'   # Contains input fastq.gz files

INPUT_PREFIX:
  - 'vector'
  - 'PANX'

INPUT_END:
  - '_1'
  - '_2' # for paired end reads

INPUT_EXT: '.fastq.gz'
  
```

- Snakefile

```
#################################### Defined by users #################################
configfile:"config/config_paired.yaml"    # Sets path to the config file

#######################################################################################

THREADS=config["THREADS"]

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all:
    input:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", 
               sample=config["INPUT_PREFIX"])



rule align_star: 
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=config["GTF"]
    output:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", sample=config['INPUT_PREFIX'])   
    params:
        indexing=config["INDEX_STAR"],
        indir=config['INPUT_DIR'],
        files=config["INPUT_PREFIX"], 
        read_ends=config['INPUT_END'],
        ext=config['INPUT_EXT']
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1='../' + params.indir + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(params.read_ends) == 2: 
                r2='../' + params.indir + params.files[i] + "_2" + params.ext + " " 
            shell("set +o pipefail; "
                  "cd star_output && " 
                  "STAR --runThreadN {THREADS} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile ../{input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix {p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions && cd ..")
        
```

- Errors: output files are created. but why do I see the "MissingOutputException" error? 

```
(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	align_star
	1	all
	2
Select jobs to execute...

[Thu Apr 15 18:21:50 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf
    output: star_output/vectorAligned.sortedByCord.out.bam, star_output/PANXAligned.sortedByCord.out.bam
    jobid: 1

Job counts:
	count	jobs
	1	align_star
	1
Apr 15 18:21:50 ..... started STAR run
Apr 15 18:21:50 ..... loading genome
Apr 15 18:22:16 ..... processing annotations GTF
Apr 15 18:22:26 ..... inserting junctions into the genome indices
Apr 15 18:23:19 ..... started 1st pass mapping
Apr 15 18:28:28 ..... finished 1st pass mapping
Apr 15 18:28:29 ..... inserting junctions into the genome indices
Apr 15 18:29:49 ..... started mapping
Apr 15 18:35:46 ..... finished mapping
Apr 15 18:35:47 ..... started sorting BAM
Apr 15 18:36:41 ..... finished successfully
Apr 15 18:36:43 ..... started STAR run
Apr 15 18:36:43 ..... loading genome
Apr 15 18:37:09 ..... processing annotations GTF
Apr 15 18:37:19 ..... inserting junctions into the genome indices
Apr 15 18:38:12 ..... started 1st pass mapping
Apr 15 18:44:01 ..... finished 1st pass mapping
Apr 15 18:44:01 ..... inserting junctions into the genome indices
Apr 15 18:45:27 ..... started mapping
Apr 15 18:52:22 ..... finished mapping
Apr 15 18:52:23 ..... started sorting BAM
Apr 15 18:53:20 ..... finished successfully
MissingOutputException in line 22 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Job Missing files after 5 seconds:
star_output/vectorAligned.sortedByCord.out.bam
star_output/PANXAligned.sortedByCord.out.bam
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 252, in handle_job_success
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/log/2021-04-15T182150.308852.snakemake.log
```


### 2. Single end 

- config/config_single.yaml

```

GTF: "reference/gencode.v36.primary_assembly.annotation.gtf" # Assigns path to .gtf file

INDEX_STAR: "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index" # Assigns path to star indexing directory (ABSOLUTE PATH NEEDED due to a potential STAR error!)

INDEX_HISAT: "reference/hisat2_index" # Assigns path to hisat2 indexing directory

THREADS: 8     # Assigns the number of threads

STAR_OUT: "star_output"  # Assigns STAR output directory

HISAT2_OUT: "hisat2_output"  # Assigns HISAT2 output directory

INPUT_DIR: 'fastq_se/'

INPUT_PREFIX:
  - 'Control'
  - 'Treatment'

INPUT_END:
  - '_1'
# - '_2' # for paired end reads

INPUT_EXT: '.fastq.gz'
  




```

- Snakefile

```

#################################### Defined by users #################################
configfile:"config/config_single.yaml"    # Sets path to the config file

#######################################################################################

THREADS=config["THREADS"]

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all:
    input:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", 
               sample=config["INPUT_PREFIX"])



rule align_star: 
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=config["GTF"]
    output:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", sample=config['INPUT_PREFIX'])   
    params:
        indexing=config["INDEX_STAR"],
        indir=config['INPUT_DIR'],
        files=config["INPUT_PREFIX"], 
        read_ends=config['INPUT_END'],
        ext=config['INPUT_EXT']
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1='../' + params.indir + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(params.read_ends) == 2: 
                r2='../' + params.indir + params.files[i] + "_2" + params.ext + " " 
            shell("set +o pipefail; "
                  "cd star_output && " 
                  "STAR --runThreadN {THREADS} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile ../{input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix {p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions && cd ..")
```

- Errors

```

(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	align_star
	1	all
	2
Select jobs to execute...

[Thu Apr 15 19:00:08 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf
    output: star_output/ControlAligned.sortedByCord.out.bam, star_output/TreatmentAligned.sortedByCord.out.bam
    jobid: 1

Job counts:
	count	jobs
	1	align_star
	1
Apr 15 19:00:08 ..... started STAR run
Apr 15 19:00:08 ..... loading genome
Apr 15 19:00:33 ..... processing annotations GTF
Apr 15 19:00:44 ..... inserting junctions into the genome indices
Apr 15 19:01:36 ..... started 1st pass mapping
Apr 15 19:02:50 ..... finished 1st pass mapping
Apr 15 19:02:51 ..... inserting junctions into the genome indices
Apr 15 19:03:53 ..... started mapping
Apr 15 19:05:21 ..... finished mapping
Apr 15 19:05:22 ..... started sorting BAM
Apr 15 19:05:39 ..... finished successfully
Apr 15 19:05:41 ..... started STAR run
Apr 15 19:05:41 ..... loading genome
Apr 15 19:06:06 ..... processing annotations GTF
Apr 15 19:06:17 ..... inserting junctions into the genome indices
Apr 15 19:07:08 ..... started 1st pass mapping
Apr 15 19:08:49 ..... finished 1st pass mapping
Apr 15 19:08:49 ..... inserting junctions into the genome indices
Apr 15 19:09:52 ..... started mapping
Apr 15 19:11:52 ..... finished mapping
Apr 15 19:11:53 ..... started sorting BAM
Apr 15 19:12:14 ..... finished successfully
MissingOutputException in line 22 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Job Missing files after 5 seconds:
star_output/ControlAligned.sortedByCord.out.bam
star_output/TreatmentAligned.sortedByCord.out.bam
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 252, in handle_job_success
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/log/2021-04-15T190007.811236.snakemake.log
```


### 3. Single end 

- config: see above

- Snakefile

```
#################################### Defined by users #################################
configfile:"config/config_single.yaml"    # Sets path to the config file

#######################################################################################

THREADS=config["THREADS"]

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all:
    input:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", 
               sample=config["INPUT_PREFIX"])



rule align_star: 
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=config["GTF"]
    output:
        "star_output/{sample}Aligned.sortedByCord.out.bam"  
    params:
        indexing=config["INDEX_STAR"],
        indir=config['INPUT_DIR'],
        files=config["INPUT_PREFIX"], 
        read_ends=config['INPUT_END'],
        ext=config['INPUT_EXT']
    run:
        shell("cd star_output")
        for i in range(len(params.files)):
            p=params.files[i]
            r1='../' + params.indir + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(params.read_ends) == 2: 
                r2='../' + params.indir + params.files[i] + "_2" + params.ext + " " 
            shell("set +o pipefail; "
                  "cd star_output && " 
                  "STAR --runThreadN {THREADS} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile ../{input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix {p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions && cd ..")
        
```

- Errors 

```

(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	2	align_star
	1	all
	3
Select jobs to execute...

[Thu Apr 15 19:20:19 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf
    output: star_output/TreatmentAligned.sortedByCord.out.bam
    jobid: 2
    wildcards: sample=Treatment

[Thu Apr 15 19:20:19 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf
    output: star_output/ControlAligned.sortedByCord.out.bam
    jobid: 1
    wildcards: sample=Control

Job counts:
	count	jobs
	1	align_star
	1
Job counts:
	count	jobs
	1	align_star
	1
Apr 15 19:20:19 ..... started STAR run
Apr 15 19:20:19 ..... loading genome
Apr 15 19:20:19 ..... started STAR run
Apr 15 19:20:19 ..... loading genome
Apr 15 19:21:00 ..... processing annotations GTF
Apr 15 19:21:00 ..... processing annotations GTF
Apr 15 19:21:11 ..... inserting junctions into the genome indices
Apr 15 19:21:11 ..... inserting junctions into the genome indices
/bin/bash: line 1: 52556 Killed                  STAR --runThreadN 8 --runMode alignReads --readFilesCommand zcat --genomeDir /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index --sjdbGTFfile ../reference/gencode.v36.primary_assembly.annotation.gtf --sjdbOverhang 100 --readFilesIn ../fastq_se/Control_1.fastq.gz --outFileNamePrefix Control --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverReadLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --outSAMunmapped None --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --twopassMode Basic --chimOutType Junctions
[Thu Apr 15 19:21:13 2021]
Error in rule align_star:
    jobid: 0
    output: star_output/ControlAligned.sortedByCord.out.bam

RuleException:
CalledProcessError in line 44 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; cd star_output && STAR --runThreadN 8 --runMode alignReads --readFilesCommand zcat --genomeDir /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index --sjdbGTFfile ../reference/gencode.v36.primary_assembly.annotation.gtf --sjdbOverhang 100 --readFilesIn ../fastq_se/Control_1.fastq.gz --outFileNamePrefix Control --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverReadLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --outSAMunmapped None --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --twopassMode Basic --chimOutType Junctions && cd ..' returned non-zero exit status 137.
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile", line 44, in __rule_align_star
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/concurrent/futures/thread.py", line 52, in run
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
Apr 15 19:22:04 ..... started 1st pass mapping
Apr 15 19:23:17 ..... finished 1st pass mapping
Apr 15 19:23:18 ..... inserting junctions into the genome indices
Apr 15 19:24:19 ..... started mapping
Apr 15 19:25:45 ..... finished mapping
Apr 15 19:25:47 ..... started sorting BAM
Apr 15 19:26:03 ..... finished successfully
Apr 15 19:26:04 ..... started STAR run
Apr 15 19:26:04 ..... loading genome
Apr 15 19:26:29 ..... processing annotations GTF
Apr 15 19:26:40 ..... inserting junctions into the genome indices
Apr 15 19:27:33 ..... started 1st pass mapping
Apr 15 19:29:10 ..... finished 1st pass mapping
Apr 15 19:29:10 ..... inserting junctions into the genome indices
Apr 15 19:30:13 ..... started mapping
Apr 15 19:32:09 ..... finished mapping
Apr 15 19:32:11 ..... started sorting BAM
 Apr 15 19:32:33 ..... finished successfully
MissingOutputException in line 22 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Job Missing files after 5 seconds:
star_output/TreatmentAligned.sortedByCord.out.bam
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 252, in handle_job_success
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/log/2021-04-15T192018.990361.snakemake.log
```


### 4. Single end 

- config: see above

- Snakefile

```

#################################### Defined by users #################################
configfile:"config/config_single.yaml"    # Sets path to the config file

#######################################################################################

THREADS=config["THREADS"]

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all:
    input:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", 
               sample=config["INPUT_PREFIX"])



rule align_star: 
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=config["GTF"]
    output:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", sample=config['INPUT_PREFIX'])  
    params:
        indexing=config["INDEX_STAR"],
        indir=config['INPUT_DIR'],
        files=config["INPUT_PREFIX"], 
        read_ends=config['INPUT_END'],
        ext=config['INPUT_EXT']
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1= params.indir + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(params.read_ends) == 2: 
                r2= params.indir + params.files[i] + "_2" + params.ext + " " 
            shell("STAR --runThreadN {THREADS} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile {input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix star_output/{p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions")
        
```

- Errors

```

(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	1	align_star
	1	all
	2
Select jobs to execute...

[Thu Apr 15 20:15:06 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf
    output: star_output/ControlAligned.sortedByCord.out.bam, star_output/TreatmentAligned.sortedByCord.out.bam
    jobid: 1

Job counts:
	count	jobs
	1	align_star
	1
Apr 15 20:15:07 ..... started STAR run
Apr 15 20:15:07 ..... loading genome
Apr 15 20:15:31 ..... processing annotations GTF
Apr 15 20:15:41 ..... inserting junctions into the genome indices
Apr 15 20:16:36 ..... started 1st pass mapping
Apr 15 20:17:49 ..... finished 1st pass mapping
Apr 15 20:17:50 ..... inserting junctions into the genome indices
Apr 15 20:18:52 ..... started mapping
Apr 15 20:20:20 ..... finished mapping
Apr 15 20:20:21 ..... started sorting BAM
Apr 15 20:20:38 ..... finished successfully
Apr 15 20:20:39 ..... started STAR run
Apr 15 20:20:39 ..... loading genome
Apr 15 20:21:04 ..... processing annotations GTF
Apr 15 20:21:15 ..... inserting junctions into the genome indices
Apr 15 20:22:08 ..... started 1st pass mapping
Apr 15 20:23:47 ..... finished 1st pass mapping
Apr 15 20:23:48 ..... inserting junctions into the genome indices
Apr 15 20:24:51 ..... started mapping
Apr 15 20:26:46 ..... finished mapping
Apr 15 20:26:47 ..... started sorting BAM
Apr 15 20:27:08 ..... finished successfully
MissingOutputException in line 19 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Job Missing files after 5 seconds:
star_output/ControlAligned.sortedByCord.out.bam
star_output/TreatmentAligned.sortedByCord.out.bam
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 252, in handle_job_success
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
```
