### 1. Single-end STAR Alignment


- Snakemake 

```


#################################### Defined by users #################################
configfile:"config/config_single.yaml"    # Sets path to the config file

#######################################################################################

THREADS=config["THREADS"]

shell.prefix('set -euo pipefail')
shell.executable('/bin/bash')

rule all:
    input:
        expand("star_output/{sample}Aligned.sortedByCord.out.bam", 
               sample=config["INPUT_PREFIX"])



rule create_dir: 
    """
    This rule creates output directories
    """
    output:
        temp("temp.txt")
    shell:
        "set +o pipefail; "
        "mkdir star_output hisat2_output && touch temp.txt"

rule align_star: 
    """
    This rule aligns the reads using STAR two-pass mode
    """
    input:
        gtf=config["GTF"],
        tempfile="temp.txt"
    output:
        "star_output/{sample}Aligned.sortedByCord.out.bam"   
    params:
        indexing=config["INDEX_STAR"],
        indir=config['INPUT_DIR'],
        files=config["INPUT_PREFIX"], 
        read_ends=config['INPUT_END'],
        ext=config['INPUT_EXT']
    run:
        for i in range(len(params.files)):
            p=params.files[i]
            r1="../" + params.indir + params.files[i] + "_1" + params.ext + " " 
            r2=""
            if len(params.read_ends) == 2: 
                r2="../"+ params.indir + params.files[i] + "_2" + params.ext + " " 
            shell("set +o pipefail; "
                  "cd star_output && " 
                  "STAR --runThreadN {THREADS} "  
                    "--runMode alignReads "  
                    "--readFilesCommand zcat "
                    "--genomeDir {params.indexing} " 
                    "--sjdbGTFfile ../{input.gtf} "  
                    "--sjdbOverhang 100 "  
                    "--readFilesIn {r1}{r2}"  
                    "--outFileNamePrefix {p} "
                    "--outFilterType BySJout "  
                    "--outFilterMultimapNmax 20 "
                    "--alignSJoverhangMin 8 "
                    "--alignSJDBoverhangMin 1 "
                    "--outFilterMismatchNmax 999 "
                    "--outFilterMismatchNoverReadLmax 0.04 "
                    "--alignIntronMin 20 "
                    "--alignIntronMax 1000000 "
                    "--outSAMunmapped None "
                    "--outSAMtype BAM "
                    "SortedByCoordinate "
                    "--quantMode GeneCounts "
                    "--twopassMode Basic "
                    "--chimOutType Junctions >> star.log && cd ..")
        
                
```


- config_single.yaml

```yaml

GTF: "reference/gencode.v36.primary_assembly.annotation.gtf" # Assigns path to .gtf file

INDEX_STAR: "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index" # Assigns path to star indexing directory (ABSOLUTE PATH NEEDED due to a potential STAR error!)

INDEX_HISAT: "reference/hisat2_index" # Assigns path to hisat2 indexing directory

THREADS: 8     # Assigns the number of threads

STAR_OUT: "star_output"  # Assigns STAR output directory

HISAT2_OUT: "hisat2_output"  # Assigns HISAT2 output directory

INPUT_DIR: 'fastq_se/'

INPUT_PREFIX:
  - 'Control'
  - 'Treatment'

INPUT_END:
  - '_1'
# - '_2' # for paired end reads

INPUT_EXT: '.fastq.gz'
  

```


- Error (non-zero exit status 137)

```

(snakemake_mapping) mira@mira-MS-7C90:~/Documents/programming/Bioinformatics/snakemake_star_hisat$ snakemake -j 8
Building DAG of jobs...
Using shell: /bin/bash
Provided cores: 8
Rules claiming more threads will be scaled down.
Job counts:
	count	jobs
	2	align_star
	1	all
	3
Select jobs to execute...

[Tue Apr 13 22:09:17 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf, temp.txt
    output: star_output/ControlAligned.sortedByCord.out.bam
    jobid: 1
    wildcards: sample=Control

[Tue Apr 13 22:09:17 2021]
rule align_star:
    input: reference/gencode.v36.primary_assembly.annotation.gtf, temp.txt
    output: star_output/TreatmentAligned.sortedByCord.out.bam
    jobid: 3
    wildcards: sample=Treatment

Job counts:
	count	jobs
	1	align_star
	1
Job counts:
	count	jobs
	1	align_star
	1
/bin/bash: line 1: 13973 Killed                  STAR --runThreadN 8 --runMode alignReads --readFilesCommand zcat --genomeDir /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index --sjdbGTFfile ../reference/gencode.v36.primary_assembly.annotation.gtf --sjdbOverhang 100 --readFilesIn ../fastq_se/Control_1.fastq.gz --outFileNamePrefix Control --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverReadLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --outSAMunmapped None --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --twopassMode Basic --chimOutType Junctions >> star.log
[Tue Apr 13 22:10:03 2021]
Error in rule align_star:
    jobid: 0
    output: star_output/TreatmentAligned.sortedByCord.out.bam

RuleException:
CalledProcessError in line 54 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Command 'set -euo pipefail;  set +o pipefail; cd star_output && STAR --runThreadN 8 --runMode alignReads --readFilesCommand zcat --genomeDir /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/reference/star_index --sjdbGTFfile ../reference/gencode.v36.primary_assembly.annotation.gtf --sjdbOverhang 100 --readFilesIn ../fastq_se/Control_1.fastq.gz --outFileNamePrefix Control --outFilterType BySJout --outFilterMultimapNmax 20 --alignSJoverhangMin 8 --alignSJDBoverhangMin 1 --outFilterMismatchNmax 999 --outFilterMismatchNoverReadLmax 0.04 --alignIntronMin 20 --alignIntronMax 1000000 --outSAMunmapped None --outSAMtype BAM SortedByCoordinate --quantMode GeneCounts --twopassMode Basic --chimOutType Junctions >> star.log && cd ..' returned non-zero exit status 137.
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 2326, in run_wrapper
  File "/home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile", line 54, in __rule_align_star
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 568, in _callback
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/concurrent/futures/thread.py", line 52, in run
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 554, in cached_or_run
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 2357, in run_wrapper
Exiting because a job execution failed. Look above for error message
MissingOutputException in line 32 of /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/Snakefile:
Job Missing files after 5 seconds:
star_output/ControlAligned.sortedByCord.out.bam
This might be due to filesystem latency. If that is the case, consider to increase the wait time with --latency-wait.
Job id: 0 completed successfully, but some output files are missing. 0
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 583, in handle_job_success
  File "/home/mira/miniconda3/envs/snakemake_mapping/lib/python3.9/site-packages/snakemake/executors/__init__.py", line 252, in handle_job_success
Exiting because a job execution failed. Look above for error message
Shutting down, this might take some time.
Exiting because a job execution failed. Look above for error message
Complete log: /home/mira/Documents/programming/Bioinformatics/snakemake_star_hisat/.snakemake/log/2021-04-13T220917.036509.snakemake.log
```


### 2. Paired-end STAR Alignment 

- Snakemake

```

```


- config_paired.yaml


```


```


- Error

```

```
